packages/contracts/contracts/web3call/combined-hooks/SRHooks.sol

```diff
@@ -0,0 +1,1099 @@
+            });
+        }
+
+        // include takedown state
+        augmentedTokenParams[originalLength + 2] = IWeb3Call.TokenParam({
+            key: "isTakedown",
+            value: tokenMetadataStorage.isTakedown ? "true" : "false"
+        });
+
+        // include send and receive states
+        // @dev do not use raw variants here, we want to be neutral if in takedown slot
+        augmentedTokenParams[originalLength + 3] = IWeb3Call.TokenParam({
+            key: "sendState",
+            value: _sendStateToString(_getSendState(tokenNumber))
+        });
+        augmentedTokenParams[originalLength + 4] = IWeb3Call.TokenParam({
+            key: "receiveState",
+            value: _receiveStateToString(_getReceiveState(tokenNumber))
+        });
+
+        // return the augmented tokenParams
+        return augmentedTokenParams;
+    }
+
+    function updateModeratorAddress(
+        address newModeratorAddress
+    ) external onlyOwner {
+        // @dev allow setting to zero address to disable moderator functionality
+        // EFFECTS
+        MODERATOR_ADDRESS = newModeratorAddress;
+        // TODO - emit event
+    }
+
+    function takedownTokenMetadataSlot(
+        uint256 tokenNumber,
+        uint256 slot
+    ) external {
+        // CHECKS
+        // msg.sender must be moderator
+        require(_isModerator(msg.sender), "Only moderator allowed");
+        // slot must be valid
+        require(slot < NUM_METADATA_SLOTS, "Invalid slot");
+        // slot must not be already takedown
+        require(
+            !tokensMetadata[tokenNumber][slot].isTakedown,
+            "Slot already takedown"
+        );
+
+        // EFFECTS
+        // takedown the token metadata slot, wiping image and sound data addresses, marking as takedown
+        TokenMetadata storage tokenMetadataStorage = tokensMetadata[
+            tokenNumber
+        ][slot];
+        tokenMetadataStorage.isTakedown = true;
+        tokenMetadataStorage.imageDataAddress = address(0); // clear the image data address
+        tokenMetadataStorage.soundDataAddress = address(0); // clear the sound data address
+        // clear the token's send/receive states, only if token is not in SendTo or ReceiveFrom state
+        // @dev SendTo and ReceiveFrom states may be expensive for owner to re-build, so we don not clear them,
+        // and instead filter on live data viewing accordingly.
+        // @dev also mitigates potential moderator abuse by removing potentially large gas usage during SendTo unwinding.
+        // send state clearing
+        // @dev effectively, this is achieved by removing the token from the general send and receive sets if it is in them
+        // and not in SendTo or ReceiveFrom state.
+        // @dev okay to optimisitically call remove - it will be a no-op if the token is not in the set
+        _sendGeneralTokens.remove(tokenNumber);
+        _receiveGeneralTokens.remove(tokenNumber);
+
+        // TODO - emit event
```

Comment:
I think this should be treated as a PMPV0 version bump, and also emit a custom "takedown" event.